flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/cache/FileBasedCache.java

```diff
@@ -123,28 +206,185 @@ public void delete(Path path) {
+        while (cacheLimitPolicy.isOverflow(value.entrySize)) {
+            moveMiddleFront();
+        }
+        cacheLimitPolicy.acquire(value.entrySize);
+    }
+
+    @Override
+    void addedToSecond(FileCacheEntry value) {
+        LOG.trace("Cache entry {} added to second link.", value.cachePath);
+        value.secondAccessEpoch = (++secondAccessEpoch);
+    }
+
+    @Override
+    void removedFromFirst(FileCacheEntry value) {
+        cacheLimitPolicy.release(value.entrySize);
+        value.close();
+    }
+
+    @Override
+    void removedFromSecond(FileCacheEntry value) {
+        value.close();
+    }
+
+    @Override
+    void movedToFirst(FileCacheEntry entry) {
+        // here we won't consider the cache limit policy.
+        // since there will be promotedToFirst called after this.
+        LOG.trace("Cache entry {} moved to first link.", entry.cachePath);
+        // trigger the loading
+        if (entry.switchStatus(
+                FileCacheEntry.EntryStatus.INVALID, FileCacheEntry.EntryStatus.LOADED)) {
+            // just a try
+            entry.loaded();
+            if (loadBackCounter != null) {
+                loadBackCounter.inc();
             }
         }
-        return value;
+        if (entry.switchStatus(
+                FileCacheEntry.EntryStatus.REMOVED, FileCacheEntry.EntryStatus.LOADING)) {
+            executorService.submit(
+                    () -> {
+                        if (entry.checkStatus(FileCacheEntry.EntryStatus.LOADING)) {
+                            Path path = entry.load();
+                            if (path == null) {
+                                entry.switchStatus(
+                                        FileCacheEntry.EntryStatus.LOADING,
+                                        FileCacheEntry.EntryStatus.REMOVED);
+                            } else if (entry.switchStatus(
+                                    FileCacheEntry.EntryStatus.LOADING,
+                                    FileCacheEntry.EntryStatus.LOADED)) {
+                                entry.loaded();
+                                if (loadBackCounter != null) {
+                                    loadBackCounter.inc();
+                                }
+                            } else {
+                                try {
+                                    path.getFileSystem().delete(path, false);
+                                    // delete the file
+                                } catch (IOException e) {
+                                }
+                            }
+                        }
+                    });
+        }
     }
 
     @Override
-    void internalInsert(String key, FileCacheEntry value) {}
+    void movedToSecond(FileCacheEntry value) {
+        // trigger the evicting
+        LOG.trace("Cache entry {} moved to second link.", value.cachePath);
+        cacheLimitPolicy.release(value.entrySize);
```

Comment:
Is the `FileCacheEntry#cachePath` deleted from  `cacheFs`?
Do we need to wait for delete before releasing `cacheLimitPolicy`?